var/spack/repos/builtin/packages/pinentry/package.py

```diff
@@ -91,17 +91,24 @@ def configure_args(self):
                 args.append("--enable-pinentry-" + gui)
             else:
                 args.append("--disable-pinentry-" + gui)
-
         return args
 
-    def test(self):
-        kwargs = {
-            "exe": self.prefix.bin.pinentry,
-            "options": ["--version"],
-            "expected": [str(self.version)],
-        }
-        self.run_test(**kwargs)
+    def check_version(self, gui):
+        if "gui=" + gui not in self.spec:
+            raise SkipTest(f"Package must be installed with {gui}")
+
+        exe = which(self.prefix.bin.pinentry + "-" + gui)
```

Comment:
I like your refactor on another PR better in that `test_pinentry` could also use `check_version` if it takes in and calls `which()` on an exe.

You'd have to move line 97-98 to within `with test_part()` and before calling `self.check_version`.  It would be nicer to set an `exe` variable before calling `test_part()` and using the `os.path.basename()` in the purpose.